pystac/extensions/file.py

```diff
@@ -220,23 +225,94 @@ def get_schema_uri(cls) -> str:
+            return cast(FileExtension[T], AssetFileExtension(obj))
+        elif isinstance(obj, pystac.Link):
+            cls.ensure_owner_has_extension(obj, add_if_missing)
+            return cast(FileExtension[T], LinkFileExtension(obj))
         else:
             raise pystac.ExtensionTypeError(cls._ext_error_message(obj))
 
 
+class AssetFileExtension(FileExtension[pystac.Asset]):
+    """A concrete implementation of :class:`FileExtension` on an
+    :class:`~pystac.Asset` that extends the Asset fields to include properties defined
+    in the :stac-ext:`File Info Extension <file>`.
+
+    This class should generally not be instantiated directly. Instead, call
+    :meth:`FileExtension.ext` on an :class:`~pystac.Asset` to extend it.
+    """
+
+    asset_href: str
+    """The ``href`` value of the :class:`~pystac.Asset` being extended."""
+
+    properties: dict[str, Any]
+    """The :class:`~pystac.Asset` fields, including extension properties."""
+
+    additional_read_properties: Iterable[dict[str, Any]] | None = None
+    """If present, this will be a list containing 1 dictionary representing the
+    properties of the owner."""
+
+    def __init__(self, asset: pystac.Asset):
+        self.asset_href = asset.href
+        self.properties = asset.extra_fields
+        if asset.owner and hasattr(asset.owner, "properties"):
+            self.additional_read_properties = [asset.owner.properties]
+
+    def __repr__(self) -> str:
+        return f"<AssetFileExtension Asset href={self.asset_href}>"
+
+
+class LinkFileExtension(FileExtension[pystac.Link]):
+    """A concrete implementation of :class:`FileExtension` on an
+    :class:`~pystac.Link` that extends the Link fields to include properties defined
+    in the :stac-ext:`File Info Extension <file>`.
+
+    This class should generally not be instantiated directly. Instead, call
+    :meth:`FileExtension.ext` on an :class:`~pystac.Link` to extend it.
+    """
+
+    link_href: str
+    """The ``href`` value of the :class:`~pystac.Link` being extended."""
+
+    properties: dict[str, Any]
+    """The :class:`~pystac.Link` fields, including extension properties."""
+
+    additional_read_properties: Iterable[dict[str, Any]] | None = None
+    """If present, this will be a list containing 1 dictionary representing the
+    properties of the owner."""
+
+    def __init__(self, link: pystac.Link):
+        self.link_href = link.href
+        self.properties = link.extra_fields
+        if link.owner and hasattr(link.owner, "properties"):
+            self.additional_read_properties = [link.owner.properties]
+
+    def __repr__(self) -> str:
+        return f"<LinkFileExtension Link href={self.link_href}>"
+
+
 class FileExtensionHooks(ExtensionHooks):
     schema_uri: str = SCHEMA_URI
-    prev_extension_ids = {"file"}
-    stac_object_types = {pystac.STACObjectType.ITEM}
+    prev_extension_ids = {
+        "file",
+        "https://stac-extensions.github.io/file/v2.0.0/schema.json",
```

Comment:
Could we also migrate from v1.0.0 as long as those removed fields aren't present? https://github.com/stac-extensions/file/blob/main/CHANGELOG.md#removed-1

Reply:
sure! We can just drop them on the floor right?

Reply:
I _think_ they can stay in the `additional_fields`, they just won't be accessible via the extension class methods? My instinct is that its better to keep them around rather than silently throwing them away, but I'd be ok w/ either way I think.

Also found a separate issue for that migration, so we could kill that too: https://github.com/stac-utils/pystac/issues/472.

It shouldn't be too many users, v2 came out _right_ after STAC v1.0.0.

Reply:
no no no. Ignore that.

Reply:
yeah I think we should do at least do a warning when updating to the new version if they have those removed fields. Then we can leave them in additional_fields. I guess I'm trying to figure out what migrate should be responsible for.

Reply:
I think warning-and-leave is the best, it's both permissive and loud :-). 👍🏼